rfc/parameterizedMetrics.md

```diff
@@ -0,0 +1,236 @@
+Parameterized Metric Support
+============================
+
+Parameterized metrics are a subset of dynamic request metrics.  Specifically they focus on modifying the calculation of metrics by altering values used in the calculation of existing metrics or mapping their values at aggregation time.
+
+The expected contract of parameterized metrics is as follows: Given a base metric with api name `metricName`, it can be suffixed with a procedure-call list of keyword parameter names and values.  These values will be used in the calculation of the resulting new metric.
+
+So
+```
+"metricName(parameterName=parameterValue)"
+```
+
+will result in a dynamic metric based on `metricName`, but substituting the parameter value when calculating it.  Also, the above string will be used as the column name in the resulting report response schema. 
+
+Motivation
+----------
+
+Standard metrics in Fili are configured at load time.  However if reporting requires many different similar caluclations, such as adjusting precision or converting values of different units into a shared aggregate, or applying custom metric filters, it may be impractical or even impossible to configure all the interesting report metrics.
+
+With parameterized metrics, open-ended variations may be supported for a given conceptual metric calculation, simplifying configuration and possibly unlocking otherwise impractical freedom in metric calculation.
+
+
+The structure of a parameterized metric:
+----------------------------------------
+
+```
+metricName(parameterName=value)
+
+metricName [ '(' parameterName=value [ ',' parameterName=value ]*  ')' ]?
+```
+
+**metricName** -> an api metric name which should appear as a resource at the `/v1/metrics/metricName` endpoint.
+
+**parameterName** -> an identifier for a parameter supported by this metric.  Identifier names should follow the same format rules as api resource identifiers.
+
+**value** -> a value to be used for creating a modified version of baseMetric.  URL encoding may be required if some of the characters in the value contain special characters.  Escaping values containing ampersand is mandatory and it may also be necessary to encode values containing comma or parentheses depending on implementation.
+
+Values containing characters delimiters such as comma or parantheses must have them nested in other scoping characters such as quotes, square brackets or parantheses.
```

Comment:
I think this is getting at "what is a valid character for a value", which is a good one to address. It also raises the question of "how to I scope my value?" In a complex case of "passing a metric to another metric", it could get messy if there are layers of parameterization. For example, how would this pseudo-metric be validly represented?

```
metricA(
    parameterOfTypeMetric=metricB(
        parameterOfTypeInt=1,
        parameterOfTypeBool=false
    ),
    parameterOfTypeMetric=metricC,
    parameterOfTypeTimeGrain=day,
    parameterOfTypeKVMap={
        key1=value1,
        key2=value2
    }
)
```

If we want parens to be the "value bound", we should indicate that in the spec. In which case, anything inside parens is treated as the "value" of the param and can't be confused by the parser. That would turn our example into this:

```
metricA(
    parameterOfTypeMetric=(
        metricB(
            parameterOfTypeInt=1,
            parameterOfTypeBool=false
        )
    ),
    parameterOfTypeMetric=metricC,
    parameterOfTypeTimeGrain=day,
    parameterOfTypeKVMap=(
        {
            key1=value1,
            key2=value2
        }
    )
)

// Compressed
metricA(parameterOfTypeMetric=(metricB(parameterOfTypeInt=1,parameterOfTypeBool=false)),parameterOfTypeMetric=metricC,parameterOfTypeTimeGrain=day,parameterOfTypeKVMap=({key1=value1,key2=value2}))
```

Reply:
I wasn't anticipating passing metrics to metrics in this fashion.
It definitely complicates parsing rules.